### PR TITLE
Move some functionality to separate files

### PR DESCRIPTION
**Pull request by eyalb181**
Move all socket management to separate file

### CODE REVIEW

File: mirrord-layer/src/sockets.rs [hunk 1]
```diff
@@ -0,0 +1,152 @@
...
+
+pub struct Sockets {
+    connections: Mutex<MultiMap<i32, u16, ConnectionSocket>>,
+    data: Mutex<HashMap<u16, DataSocket>>,
+    connection_queue: Mutex<Queue<u16>>,
+    pending_data: Mutex<HashMap<u16, Vec<u8>>>
+}
+
+impl Sockets {
+    pub fn new() -> Self {
```
Comment (meowjesty) [inline mirrord-layer/src/sockets.rs:36]: Minor thing, if `new` takes no arguments, then it should probably be an implementation of `default` instead.

```suggestion
impl Default for Sockets {
    fn default() -> Self { ... }
}
```

**Review by meowjesty — COMMENTED** [review]
👍

File: mirrord-layer/src/sockets.rs [hunk 2]
```diff
@@ -0,0 +1,152 @@
...
+    pub write_socket: SocketpairStream,
+}
+
+pub struct Sockets {
+    connections: Mutex<MultiMap<i32, u16, ConnectionSocket>>,
+    data: Mutex<HashMap<u16, DataSocket>>,
+    connection_queue: Mutex<Queue<u16>>,
+    pending_data: Mutex<HashMap<u16, Vec<u8>>>
+}
+
```
Comment (aviramha) [inline mirrord-layer/src/sockets.rs:34]: perhaps add some docs here

**Review by aviramha — CHANGES_REQUESTED** [review]

File: mirrord-layer/src/lib.rs [hunk 3]
```diff
@@ -89,47 +58,16 @@ unsafe extern "C" fn bind_detour(sockfd: i32, addr: *const sockaddr, addrlen: so
...
-    let real_fd = match CONNECTION_SOCKETS.lock().unwrap().get(&sockfd) {
-        Some(s) => s.real_read_socket_fd,
-        None => {
-            println!("No socket found for fd: {}", sockfd);
-            return -1;
-        }
-    };
-
-    libc::connect(real_fd, addr, addrlen)
+    SOCKETS.add_port_to_connection_socket(sockfd, port);
```
[the remark this answers is not in the document]
Comment (eyalb181) [reply mirrord-layer/src/lib.rs:61]: I had some theoretical logic there (where I cached the real socket) but I'd rather run into real behaviour and then decide what to do.

File: mirrord-layer/src/lib.rs [hunk 4]
```diff
@@ -14,72 +10,45 @@ use std::{
...
+    let (sender, receiver) = channel::<i32>(1000);
+    *NEW_CONNECTION_SENDER.lock().unwrap() = Some(sender);
+    enable_hooks();
+    RUNTIME.spawn(poll_agent(pf, receiver));
+}
 
-    read_fd
+unsafe extern "C" fn socket_detour(_domain: i32, _socket_type: i32, _protocol: i32) -> i32 {
+    debug!("socket called");
+    SOCKETS.create_connection_socket()
```
[the remark this answers is not in the document]
Comment (eyalb181) [reply mirrord-layer/src/lib.rs:51]: But it doesn't just create the socket, it also adds it to the connection sockets map, and uses it internally later

File: mirrord-layer/src/lib.rs [hunk 5]
```diff
@@ -172,138 +110,22 @@ unsafe extern "C" fn accept_detour(
         let os_addr: OsSocketAddr = socket_addr.into();
```
[the remark this answers is not in the document]
Comment (eyalb181) [reply mirrord-layer/src/lib.rs:110]: Implement what?

File: mirrord-layer/src/pod_api.rs [hunk 6]
```diff
@@ -0,0 +1,93 @@
...
+                        }
+                    ],
+                    "command": [
+                        "./mirrord-agent",
+                        "--container-id",
+                        container_id,
+                        "-t",
+                        "60"
+                    ],
+                    "env": [{"name": "RUST_LOG", "value": "trace"}],
```
Comment (aviramha) [inline mirrord-layer/src/pod_api.rs:75]: RUST_LOG should be env controlled

File: mirrord-layer/src/sockets.rs [hunk 7]
```diff
@@ -0,0 +1,222 @@
...
+#[derive(PartialEq)]
+pub enum ConnectionState {
+    Bound,
+    Listening,
+}
+
+pub struct DataSocket {
+    pub connection_id: u16,
+    #[allow(dead_code)]
+    pub read_socket: SocketpairStream, /* Though this is unread, it's necessary to keep the
```
Comment (aviramha) [inline mirrord-layer/src/sockets.rs:37]: We actually should read it so buffers wont fill.

File: mirrord-layer/src/sockets.rs [hunk 8]
```diff
@@ -0,0 +1,222 @@
...
+                                        * socket open */
+    pub write_socket: SocketpairStream,
+    pub address: SocketAddr,
+}
+
+pub struct Sockets {
+    new_sockets: Mutex<HashMap<i32, Socket>>,
+    connections: Mutex<MultiMap<i32, u16, ConnectionSocket>>, // Keys are sockfd, port
```
Comment (aviramha) [inline mirrord-layer/src/sockets.rs:45]: You can define types so code will be more readable even without docs - 
`type SockFd = i32` `type Port = u16`

**Review by aviramha — CHANGES_REQUESTED** [review]
Very good job, few last comments ;)

File: mirrord-layer/src/sockets.rs [hunk 9]
```diff
@@ -0,0 +1,240 @@
...
+            // NewTcpConnection
+            debug!("No socket found for connection_id: {}", connection_id);
+            self.pending_data
+                .lock()
+                .unwrap()
+                .insert(connection_id, data);
+        }
+    }
+
+    fn clear_data(&self, socket: &mut SocketpairStream) {
```
Comment (aviramha) [inline mirrord-layer/src/sockets.rs:224]: no use of self, make this a module level function

**Review by aviramha — APPROVED** [review]

**Pull request merged**